### ISSUE TITLE
[RLLib] NCCL rank desync timeout with num_learners=4 and MultiRLModule containing 2 heterogeneous policies (multi-agent PPO)

### ISSUE THREAD

**Issue by alvaro-rio** [ray-project/ray#63819]
### What happened + What you expected to happen

When using `num_learners=4` with a `MultiRLModule` containing **two heterogeneous policies** 
(different observation sizes and different numbers of agent-steps per batch), NCCL collective 
operations desync across ranks after ~30 minutes, crashing with a watchdog timeout.

The same setup with `num_learners=4` and a **single policy** (same hardware, same config) works 
correctly. Reducing to `num_learners=2` with two policies also works correctly.

### Versions / Dependencies

- Ray version: 2.47.0
- PyTorch version: 2.7.0+cu126
- Python: 3.12.3
- GPU: 4× NVIDIA L40S (no NVLink, PCIe connected)
- OS: Ubuntu (Linux)

### Reproduction script

import numpy as np
import torch
import torch.nn as nn
from typing import Any

import ray
from ray import tune
from ray.tune import RunConfig
from ray.rllib.env.multi_agent_env import MultiAgentEnv
from ray.rllib.algorithms.ppo import PPOConfig
from ray.rllib.policy.policy import PolicySpec
from ray.rllib.core.rl_module.rl_module import RLModuleSpec
from ray.rllib.core.rl_module.multi_rl_module import MultiRLModuleSpec
from ray.rllib.core.rl_module.torch import TorchRLModule
from ray.rllib.core.rl_module.apis.value_function_api import ValueFunctionAPI
from ray.rllib.core.columns import Columns
from ray.rllib.utils.annotations import override
from gymnasium import spaces
import gymnasium as gym

# 4 agents use policy_a, 2 agents use policy_b
AGENTS_A = ["a0", "a1", "a2", "a3"]
AGENTS_B = ["b0", "b1"]
ALL_AGENTS = AGENTS_A + AGENTS_B
OBS_A, OBS_B = 143, 139

class DummyMAEnv(MultiAgentEnv):
    def __init__(self, config=None):
        super().__init__()
        self._agent_ids = set(ALL_AGENTS)
        self._step = 0

    def reset(self, *, seed=None, options=None):
        self._step = 0
        return self._obs(), {}

    def step(self, action_dict):
        self._step += 1
        done = self._step >= 200
        t  = {a: done  for a in ALL_AGENTS}; t["__all__"]  = done
        tr = {a: False for a in ALL_AGENTS}; tr["__all__"] = False
        return self._obs(), {a: 0.0 for a in ALL_AGENTS}, t, tr, {}

    def _obs(self):
        return {
            a: np.random.uniform(-1, 1,
                OBS_A if a in AGENTS_A else OBS_B
            ).astype(np.float32)
            for a in ALL_AGENTS
        }

class SimpleLSTM(TorchRLModule, ValueFunctionAPI):
    @override(TorchRLModule)
    def setup(self):
        obs  = self.model_config["obs_size"]
        act  = self.model_config["act_size"]
        self.hidden = 128
        self.enc  = nn.Sequential(nn.Linear(obs, self.hidden), nn.ReLU())
        self.lstm = nn.LSTM(self.hidden, self.hidden, batch_first=True)
        self.mean = nn.Linear(self.hidden, act)
        self.lstd = nn.Linear(self.hidden, act)
        self.vf   = nn.Sequential(nn.Linear(obs, self.hidden), nn.ReLU(), nn.Linear(self.hidden, 1))

    @override(TorchRLModule)
    def get_initial_state(self) -> Any:
        return {"h": np.zeros((self.hidden,), np.float32),
                "c": np.zeros((self.hidden,), np.float32)}

    def _fwd(self, batch):
        obs = batch[Columns.OBS]
        if obs.dim() == 2: obs = obs.unsqueeze(1)
        B, T, _ = obs.shape
        x = self.enc(obs.reshape(B * T, -1)).reshape(B, T, -1)
        sin = batch.get(Columns.STATE_IN)
        h = sin["h"].unsqueeze(0) if sin else torch.zeros(1, B, self.hidden, device=obs.device)
        c = sin["c"].unsqueeze(0) if sin else torch.zeros(1, B, self.hidden, device=obs.device)
        out, (h, c) = self.lstm(x, (h.to(obs.device), c.to(obs.device)))
        return {
            Columns.ACTION_DIST_INPUTS: torch.cat([self.mean(out), self.lstd(out).clamp(-3, -0.5)], -1),
            Columns.STATE_OUT: {"h": h.squeeze(0), "c": c.squeeze(0)},
            Columns.EMBEDDINGS: out,
        }

    @override(TorchRLModule)
    def _forward_inference(self, batch, **kw): return self._fwd(batch)
    @override(TorchRLModule)
    def _forward_train(self, batch, **kw): return self._fwd(batch)

    @override(ValueFunctionAPI)
    def compute_values(self, batch, embeddings=None):
        obs = batch[Columns.OBS]
        if obs.dim() == 2: obs = obs.unsqueeze(1)
        B, T, _ = obs.shape
        return self.vf(obs.reshape(B * T, -1)).reshape(B, T)

if __name__ == "__main__":
    ray.init()

    cfg = (
        PPOConfig()
        .environment(DummyMAEnv, disable_env_checking=True)
        .framework("torch")
        .training(train_batch_size=40_000, minibatch_size=10_000, num_epochs=10, lr=3e-4)
        .multi_agent(
            policies={
                "policy_a": PolicySpec(
                    observation_space=spaces.Box(-1, 1, (OBS_A,), dtype=np.float32),
                    action_space=spaces.Box(-1, 1, (4,), dtype=np.float32),
                ),
                "policy_b": PolicySpec(
                    observation_space=spaces.Box(-1, 1, (OBS_B,), dtype=np.float32),
                    action_space=spaces.Box(-1, 1, (2,), dtype=np.float32),
                ),
            },
            policy_mapping_fn=lambda aid, *a, **kw: "policy_a" if aid in AGENTS_A else "policy_b",
            policies_to_train=["policy_a", "policy_b"],
        )
        .rl_module(
            model_config={"max_seq_len": 50},
            rl_module_spec=MultiRLModuleSpec(rl_module_specs={
                "policy_a": RLModuleSpec(SimpleLSTM, model_config={"obs_size": OBS_A, "act_size": 4}),
                "policy_b": RLModuleSpec(SimpleLSTM, model_config={"obs_size": OBS_B, "act_size": 2}),
            }),
        )
        .env_runners(num_env_runners=4, num_envs_per_env_runner=4,
                     batch_mode="truncate_episodes", rollout_fragment_length="auto")
        # BUG: hangs ~30min then crashes with NCCL watchdog timeout
        .learners(num_learners=4, num_gpus_per_learner=1.0)
        .resources(num_cpus_for_main_process=1, num_gpus=4.0)
        # WORKAROUND: num_learners=2 avoids the NCCL desync
        # .learners(num_learners=2, num_gpus_per_learner=1.0)
        # .resources(num_cpus_for_main_process=1, num_gpus=2.0)
    )

    tune.Tuner(
        "PPO",
        param_space=cfg.to_dict(),
        run_config=RunConfig(stop={"training_iteration": 5}, verbose=1),
    ).fit()
    ray.shutdown()

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by tobby168** [comment]
I've been digging into this, and the key finding is that **it reproduces on CPU with the GLOO backend — no GPUs needed to repro or test.**

**Root cause.** With `num_learners > 1`, RLlib wraps *each* submodule of a `MultiRLModule` in its own `TorchDDPRLModule` (`torch_learner.py::_make_modules_ddp_if_necessary`, gated on `num_learners`, not GPU). A learner only runs `forward_train`/backward for modules present in its data shard — `MultiRLModule._forward_train` iterates `for mid in batch.keys() if mid in self`, `compute_losses` builds losses only `for module_id in fwd_out`, and `compute_gradients` does `total_loss = sum(loss_per_module.values()); total_loss.backward()`. So when a shard is missing a policy (which `ShardEpisodesIterator` can produce, especially for the minority policy at higher `num_learners`), that learner never enters that policy's DDP all-reduce while the learners that *do* have it block on the collective → NCCL watchdog timeout (GLOO timeout on CPU). This lines up with num_learners=4-fails / =2-works and single-policy-works.

**Minimal CPU repro (no Ray, no GPU — `pip install torch`):**
```python
import os, socket, sys
from datetime import timedelta
import torch, torch.distributed as dist, torch.multiprocessing as mp, torch.nn as nn
from torch.nn.parallel import DistributedDataParallel as DDP

def worker(rank, mode):
    dist.init_process_group("gloo", init_method="env://", rank=rank, world_size=2,
                            timeout=timedelta(seconds=8))
    policy_a, policy_b = DDP(nn.Linear(4, 4)), DDP(nn.Linear(4, 4))
    x = torch.randn(2, 4)
    loss = policy_a(x).sum()
    # rank 1's shard is "missing" policy_b -> it skips policy_b's backward (the bug)
    if mode == "control" or rank == 0:
        loss = loss + policy_b(x).sum()
    try:
        loss.backward(); dist.barrier(); print(f"[rank {rank}] DONE")
    except Exception as e:
        print(f"[rank {rank}] DESYNC: {str(e).splitlines()[0]}")
    dist.destroy_process_group()

if __name__ == "__main__":
    s = socket.socket(); s.bind(("", 0)); port = s.getsockname()[1]; s.close()
    os.environ["MASTER_ADDR"], os.environ["MASTER_PORT"] = "127.0.0.1", str(port)
    mp.spawn(worker, args=(sys.argv[1] if len(sys.argv) > 1 else "buggy",), nprocs=2)
```
`python repro.py control` → both ranks `DONE`; `python repro.py buggy` → `[gloo] Timed out waiting ... for recv` (the CPU analogue of the reported NCCL timeout).

**Fix direction — I tested a few approaches on the repro:**
- Adding a zero term over the absent module's *parameters* (no forward) does **not** work — DDP arms its reducer during `forward()`, so the all-reduce still never fires → still deadlocks.
- **(A) Participate via a zero-scaled forward:** run `forward_train` through the absent module and add `0 * fwd_out` to the loss so its all-reduce fires in lockstep. Uses all data, but averages that module's gradient over `world_size` (diluting a minority policy present on only some ranks) and needs a valid dummy input (incl. state for RNNs).
- **(B) Intersection:** all-gather each rank's present module IDs and process only the intersection (modules present on every rank this minibatch). No gradient dilution and no dummy inputs, but drops a module's data in minibatches where some rank lacks it.

Both resolve the deadlock in the repro. I'd be happy to pick this up and implement it with a regression test — **is it OK if I take this one?** And which direction do you prefer (A vs B), or is there a pattern you'd rather use (e.g. a single DDP over the whole `MultiRLModule` with `find_unused_parameters=True`)?

**Comment by tobby168** [comment]
One more design angle, since the deadlock is really a symptom of the sharding not guaranteeing per-policy coverage — it's worth deciding **which layer should own the "every rank touches every module each step" invariant**: the data layer (sharding) or the sync layer (DDP participation).

A couple of observations on the data-layer side:

- **`ShardBatchIterator`** slices each policy with `math.ceil(len(sub_batch) / num_shards)`, which empties the trailing shard(s) whenever a policy has fewer (or not-evenly-divisible) samples than `num_shards` (e.g. 6 samples / 4 shards → `[2,2,2,0]`). A balanced split (`np.array_split`-style `[2,2,1,1]`) would guarantee coverage **as long as the policy has ≥ num_shards samples**, and is a cheap, standalone data-efficiency fix regardless of the deadlock.
- **`ShardEpisodesIterator`** (the path this issue hits, since it's stateful/LSTM → episodes) only balances total env-steps and shards whole episodes, so per-policy coverage is incidental. Guaranteeing coverage there is a multi-objective bin-packing problem, and episodes can't be sliced to redistribute a policy's timesteps.

The fundamental limit: when a minority policy has **fewer samples/episodes than `num_learners` in a given batch**, no sharding can give every learner ≥1 sample without duplicating data — so sharding alone can't fully guarantee the invariant (and the minibatch iterator can drop a policy even when the shard has it).

So this looks like a defense-in-depth situation: improve the split so the mismatch rarely happens (and never for the common case), **and** keep a lockstep-safe sync fallback (e.g. the "intersection" approach) for the residual case. Happy to implement whichever split/sync combination you prefer — mainly wanted to flag that a better split helps but can't be the whole fix. Which layer would you rather own the invariant?

**Comment by ArturNiederfahrenhorst** [comment]
Can you explain in what situation a TorchLearner would not have an instance of each RLModule/Policy ?

**Comment by tobby168** [comment]
@ArturNiederfahrenhorst — following up on your question (and my own) — I reproduced this on **CPU with the gloo backend** (`num_learners=4`, `num_gpus=0`, built-in default RLModule + `use_lstm`, two heterogeneous policies) and traced it end-to-end. Short version: **you're right that every learner has an instance of every policy — the deadlock is not a missing module. It's a per-module *ordering* divergence across learners.**

The differential matches the report exactly:

| config | result |
|---|---|
| 1 policy, `num_learners=4` | completes |
| 2 policies, `num_learners=2` | completes |
| 2 policies, `num_learners=4` | crashes on the first collective |

(single-policy at the same `world_size=4` completing rules out generic gloo flakiness.)

Tracing all four learners right before the crash: every one has **both** policies present, `num_total_minibatches=20`, and identical per-module sizes — so no missing module and no unequal minibatch count. But one learner's `batch.policy_batches` is ordered `[policy_b, policy_a]` while the other three are `[policy_a, policy_b]`:

```
learner 0/1/2:  lens={'policy_a': 2052, 'policy_b': 1026}
learner 3:      lens={'policy_b': 1026, 'policy_a': 2052}   <-- opposite order
```

Since each submodule is wrapped in its own `TorchDDPRLModule` and `MultiRLModule._forward_train` iterates `for mid in batch.keys()`, the odd learner runs forward/backward on the two modules in the opposite order → its two per-submodule gradient all-reduces are issued in the opposite sequence. `policy_a` and `policy_b` have different parameter sizes, so at the same collective step one rank all-reduces a `policy_a`-sized tensor while the others all-reduce a `policy_b`-sized one → gloo aborts with a size mismatch (`enforce fail … op.nread == op.preamble.nbytes`). On NCCL/GPU that same mismatch just blocks → the 30-min watchdog timeout you hit. It also explains the knobs: single policy → only one all-reduce, nothing to mis-order; fewer learners → smaller chance any rank's order diverges.

**Fix (verified):** make the per-module iteration order deterministic and identical on every learner — iterate `sorted(batch.keys())` (or otherwise fix `policy_batches` ordering at construction) in the module loops that feed DDP (`MultiRLModule._forward_train`, and the `compute_losses`/`compute_gradients` loops). With that one change the `num_learners=4` two-policy repro goes from crashing **3/3** runs to passing **3/3**.

This supersedes my earlier "`ShardEpisodesIterator` drops the minority policy" theory for this repro — that genuinely can't happen here, since every episode carries both policies, so every shard does too. A learner only ends up *without* a policy's data when that policy's agents aren't in every episode, or via `ShardBatchIterator`'s `ceil()` on the non-episode path — a separate latent issue worth hardening (lockstep/intersection fallback), but not what bites this repro.

Happy to open a PR with the deterministic-ordering fix plus a CPU/gloo regression test (`num_learners=4`, two heterogeneous policies) that fails on master and passes with the fix. Want me to go ahead?
